_posts/2024-12-02-safety-of-methods-for-numeric-primitive-types.md

```diff
@@ -0,0 +1,195 @@
+
+In the past 3 months, we have rigorously analyzed unsafe methods provided by Rust's numeric primitive types, such as `unchecked_add` and `unchecked_sub`, which omit runtime checks for overflow and underflow, using formal verification techniques.
+
+## Challenge Overview
+The [challenge](https://model-checking.github.io/verify-rust-std/challenges/0011-floats-ints.html) is divided into three parts:
+- [Part 1] Unsafe Integer Methods: Prove safety of methods like ```unchecked_add```, ```unchecked_sub```, ```unchecked_mul```, ```unchecked_shl```, ```unchecked_shr```, and ```unchecked_neg``` for various integer types.
+- [Part 2] Safe API Verification: Verify safe APIs that leverage the unsafe integer methods from Part 1, such as ```wrapping_shl```, ```wrapping_shr```, ```widening_mul```, and ```carrying_mul```.
+- [Part 3] Float to Integer Conversion: Verify the absence of undefined behavior in ```to_int_unchecked``` for floating-point types ```f32``` and ```f64```. **FIXME: should we include f16 and f128 as well?**
+
+In addition to verifying the methods under their specified safety preconditions, we needed to ensure the absence of specific undefined behaviors:
+- Invoking undefined behavior via compiler intrinsics.
+- Reading from uninitialized memory.
+- Mutating immutable bytes.
+- Producing an invalid value.
+
+## Approach and Implementation
+To tackle this challenge, we utilized Kani's verification capabilities to write proof harnesses for the unsafe methods. Our strategy involved:
+1. **Specifying Safety Preconditions**: To ensure the methods behaved correctly, we explicitly specified safety preconditions in the code. These preconditions were used to:
+   - Guide input generation.
+   - Define the boundaries for valid inputs.
+   - Avoid triggering undefined behavior in cases where the preconditions were violated.
+
+2. **Writing Verification Harnesses**: We developed Kani proof harnesses for each unsafe method by:
+   - Generating inputs that satisfy the required safety preconditions.
+   - Invoking the methods within an `unsafe` block to verify their correctness and safety under the defined conditions.
+
+3. **Using Macros for Code Generation**: To efficiently handle the large number of methods and types that required verification, we utilized Rust macros. These macros automated the generation of proof harnesses, reducing redundancy and improving maintainability.
+
+4. **Handling Large Input Spaces**: For methods with large input spaces (e.g., `unchecked_mul` for 64-bit integers), we made the verification process more tractable by:
+   - Introducing assumptions to limit the range of inputs.
+   - Focusing on specific edge cases (e.g., maximum values, zero, and boundary conditions) that are most likely to cause overflow or underflow.
+
+We will walk through each part with examples from the Unsafe Integer Methods part of the challenge.
+
+### 1. Specifying Safety Preconditions
+```rust
```

Comment:
I think it would make for slightly nicer reading if there were one or two sentences of text between the heading and the example. Such text should explain what the source code that follows will exemplify.

Reply:
fixed